pennylane/transforms/mitigate.py

```diff
@@ -12,15 +12,166 @@
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 2)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──RX(3.0)†──RX(3.0)──────────────────┤ ╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)───RY(4.0)†─╭●──╭●──RY(4.0)─┤ ├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†─╰X──RZ(5.0)─┤ ╰<Z@Z@Z>
+
+    Setting ``scale_factor = 3`` results in the folded circuit :math:`U (U^\dagger U)`.
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 3)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──RX(3.0)†───────────────╭●─────────RX(0.0)†──RX(0.0)─╭●──RX(3.0)──────────┤╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)───RY(4.0)†─╭●──╰X†────────RY(1.0)†──RY(1.0)─╰X─╭●────────RY(4.0)─┤├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†──RZ(2.0)†──RZ(2.0)──────────────╰X────────RZ(5.0)─┤╰<Z@Z@Z>
+
+    .. note::
+
+        Circuits are treated as lists of operations. Since the ordering is ambiguous, as seen exemplarily
+        for :math:`U = X(0) Y(0) X(1) Y(1) = X(0) X(1) Y(0) Y(1)`, also partially folded circuits are ambiguous.
+    """
+
+    if scale_factor < 1.0:
+        raise AttributeError("scale_factor must be >= 1")
+    assert scale_factor >= 1.0
+
+    # TODO: simplify this - currently just a workaround, problem is when tape contains adjoint(op)
+    def qfunc(op):
+        copy(op).queue()
+
+    # Generate base_circuit without measurements
+    # Treat all circuits as lists of operations, build new tape in the end
+    base_ops = circuit.expand().copy(copy_operations=True).operations
+
+    def _divmod(a, b):
+        """Performs divmod but in an all-interface compatible manner"""
+        out1 = qml.math.floor(a / b)
+        out2 = a - out1 * b
+        return int(out1), int(out2)
+
+    num_global_folds, fraction_scale = _divmod(scale_factor - 1, 2)
+
+    n_ops = len(base_ops)
+    num_to_fold = int(round(fraction_scale * n_ops / 2))
 
-# pylint: disable=too-many-arguments, protected-access
+    # Create new_circuit from folded list
+    with QuantumTape() as new_circuit:
+        # Original U
+        for op in base_ops:
+            qfunc(op)
+
+        # Folding U => U (U^H U)**n.
+        for _ in range(int(num_global_folds)):
+            for op in base_ops[::-1]:
+                adjoint(qfunc)(op)
+
+            for op in base_ops:
+                qfunc(op)
+
+        # Remainder folding U => U (U^H U)**n (L_d^H .. L_s^H) (L_s .. L_d)
+        for i in range(n_ops - 1, n_ops - num_to_fold - 1, -1):
+            adjoint(qfunc)(base_ops[i])
+
+        for i in range(n_ops - num_to_fold, n_ops):
+            qfunc(base_ops[i])
+
+        # Append measurements
+        for meas in circuit.measurements:
+            apply(meas)
+
+    return new_circuit
+
+
+# TODO: make this a pennylane.math function
+def _polyfit(x, y, order):
+    """Brute force implementation of polynomial fit"""
+    lhs = qml.math.vander(x, order + 1)
```

Comment:
I want to somehow ensure that x is automatically casted into the same dtype as y (which is determined by the interface) so the user does not need to worry about the format of the scale_factors. At the moment the user has to pass the scale_factors in the right format otherwise errors occur.

Reply:
this is still open

Reply:
could you do?
```python
x = qml.math.cast_like(x, y)
```
(see https://pennylane--2757.org.readthedocs.build/en/2757/code/api/pennylane.math.cast_like.html)

Reply:
Yes this is exactly what I was looking for, but adding this line breaks the code, not just for jax-jit but also for autograd, jax and torch. Only TF seems unfazed by the change.

Reply:
@rmoyard found a solution 👼